support/export/rmtab.c

```diff
@@ -96,3 +97,98 @@ rmtab_read(void)
 	}
 	return 0;
 }
+
+/*
+ * Check if an rmtab_path is exported given an export_path from etab file.
+ * The path is considered exported if it is an exact match or is part of
+ * a path nested under an export.
+ */
+static int
+is_exported(const char* rmtab_path, const char* export_path) {
+	/* first check that rmtab_path starts with the export_path */
+	if (strstr(rmtab_path, export_path) != rmtab_path)
+		return 0;
+
+	/* then confirm that it is an exact match or that it's nested */
+	char next = rmtab_path[strlen(export_path)];
+	return next == '\0' || next == '/';
+}
+
+/*
+ * The rmtab file can acculmulate stale entries over time. So when the list of
+ * exports changes, make a best effort attempt to insure that every entry in
+ * rmtab has a corresonding export.
+ */
+void
+rmtab_rebuild(void)
+{
+	struct rmtabent	*rep;
+	struct stat	stb;
+	FILE		*fp;
+	int		rmtab_lockid, etab_lockid;
+	int		entries_seen = 0, entries_copied = 0;
+
+	/* Hold locks for the rmtab and etab files during rebuilding */
+	if ((rmtab_lockid = xflock(_PATH_RMTABLCK, "w")) < 0)
+		return;
+
+	/* fast path for no NFSv3 mounts */
+	if (stat(_PATH_RMTAB, &stb) < 0 || stb.st_size == 0) {
+		xfunlock(rmtab_lockid);
+		return;
+	}
+
+	if ((etab_lockid = xflock(_PATH_ETABLCK, "r")) < 0) {
+		xfunlock(rmtab_lockid);
+		return;
+	}
+
+	if (!setrmtabent("r")) {
+		xfunlock(rmtab_lockid);
+		xfunlock(etab_lockid);
+		return;
+	}
+	if (!(fp = fsetrmtabent(_PATH_RMTABTMP, "w"))) {
+		endrmtabent();
+		xfunlock(rmtab_lockid);
+		xfunlock(etab_lockid);
+		return;
+	}
+
+	/* Visit all the entries in rmtab file */
+	while ((rep = getrmtabent(1, NULL)) != NULL) {
+		struct exportent *xp;
+		int saved = 0;
+
+		/* filter entries that are not exported */
+		setexportent(_PATH_ETAB, "r");
+		while ((xp = getexportent_nofollow(0, 0)) != NULL) {
```

Comment:
Could this O(n^2) operation ever get expensive? How big can these files get?

Reply:
The size of the `etab` file is base on the number of exported NFS filesystems (call this N). We don't expect this to be > 1,000 on most customer systems.  On `dcol2`, `etab` has 1,766 entries and a size of 385K.  The number of entries in `rmtab` tends to be smaller than the total exports (call this M).

So we expect to see M opens of the `etab` file and for each opened file see N lines processed using `getc()` (which on Linux will be doing 8K reads) and these stream reads will be coming from the file system cache.

I'll measure this added cost using a copy of the `dcol2`, `etab` file...

Reply:
On a T3_2XLARGE system with 1,556 `rmtab` entries and 1,614 `etab` entries it takes just over 8 seconds to rebuild the rmtab file.
```
delphix@ip-10-110-200-69:~$ wc -l /var/lib/nfs/rmtab
1556 /var/lib/nfs/rmtab
delphix@ip-10-110-200-69:~$ wc -l /var/lib/nfs/etab
1614 /var/lib/nfs/etab
delphix@ip-10-110-200-69:~$ time sudo ./exportfs-rebuild -ra

real	0m8.276s
user	0m8.115s
sys	0m0.156s
```

Reply:
I refactored `rmtab_rebuild` to build a list of exported paths to avoid constantly re-opening and re-reading etab file.  The rebuild now takes a fraction of a second.